syncer/internal/controller/syncer.go

```
@@ -0,0 +1,162 @@
+package controller
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/awslabs/operatorpkg/context"
+	clsv1 "github.com/llmariner/cluster-manager/api/v1"
+	"golang.org/x/sync/errgroup"
+	batchv1 "k8s.io/api/batch/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	metricsserver "sigs.k8s.io/controller-runtime/pkg/metrics/server"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// RemoteSyncerManager manages remote syncers.
+type RemoteSyncerManager struct {
+	sessionManagerEndpoint string
+
+	localK8sClient client.Client
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (m *RemoteSyncerManager) SetupWithManager(mgr ctrl.Manager, sessionManagerServerAddr string) error {
+	m.sessionManagerEndpoint = sessionManagerServerAddr
+	m.localK8sClient = mgr.GetClient()
+	return mgr.Add(m)
+}
+
+// Start starts the remote syncer manager.
+func (m *RemoteSyncerManager) Start(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx).WithName("syncer")
+	log.Info("Starting remote syncer manager")
+
+	// TODO(aya): support authentication & dynamic cluster registration
+	const token = "dummy"
+	cls := []*clsv1.Cluster{{Id: "default"}}
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	for _, c := range cls {
+		log.Info("Starting remote syncer", "clusterName", c.Name)
+		eg.Go(func() error {
+			ctx := ctrl.LoggerInto(egCtx, log.WithName(c.Id))
+			rconf := getRestConfig(m.sessionManagerEndpoint, c.Id, token)
+
+			// TODO(aya): gracefully handle errors
+			if err := newStatusSyncer(m.localK8sClient).start(ctx, rconf); err != nil {
+				return fmt.Errorf("run status syncer %s(%s): %s", c.Name, c.Id, err)
+			}
+			return nil
+		})
+	}
+
+	if err := eg.Wait(); err != nil {
```

`eg.Wait()` waits until all goroutines complete, but I think we want to return an error once one of them completes. 

https://pkg.go.dev/golang.org/x/sync/errgroup#Group.Wait

If this makes sense, could you add a TODO comment?

---

This error group is created with a context in L44. The contexts are canceled when the group receives a first error, and the status syner uses that context.